## Домашнее задание 1 - 10 баллов

1. Загрузите набор данных lenta-ru-news с помощью библиотеки Corus для задачи классификации текстов по топикам (пригодятся атрибуты title, text, topic)- **1 балл**
2. Подготовьте данные к обучению: - **3 балла**
    - Предобработайте данные: реализуйте оптимальную, на ваш взгляд, предобработку текстов (нормализация, очистка, стемминг/лемматизация и т.п.) и таргета.
    - **hint**: для ускорения обработки  и обучения можно ограничиться не всем датасетом, а его репрезентативной частью, например, размера 100_000.
    - Кратко опишите пайплайн, на котором остановились, и почему.
    - Разделите датасет на обучающую, валидационную и тестовую выборки со стратификацией в пропорции 60/20/20. В качестве целевой переменной используйте атрибут `topic`
3. Замерьте базовое качество с любым dummy-бейзлайном - **0.5 балла**
4. Обучите модель `sklearn.linear_model.LogisticRegression` с двумя вариантами векторизации: **2 балла**
  - `sklearn.feature_extraction.text.CountVectorizer`
  - `sklearn.feature_extraction.text.TfidfVectorizer`
5. Попробуйте улучшить качество, подобрав оптимальные гиперпараметры трансформаций и модели на кросс-валидации **1 балл**
6. Оцените качество лучшего пайплайна на отложенной выборке - **0.5 балла**

**Задание выполнялось в Google Colab**

In [1]:
!wget https://github.com/yutkin/Lenta.Ru-News-Dataset/releases/download/v1.0/lenta-ru-news.csv.gz

--2025-03-03 11:47:05--  https://github.com/yutkin/Lenta.Ru-News-Dataset/releases/download/v1.0/lenta-ru-news.csv.gz
Resolving github.com (github.com)... 140.82.112.4
Connecting to github.com (github.com)|140.82.112.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/87156914/0b363e00-0126-11e9-9e3c-e8c235463bd6?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20250303%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20250303T114705Z&X-Amz-Expires=300&X-Amz-Signature=ac0f673f48082abb5e60f8de23f575c82b72b3590f8f27282cbaf9693ed36c5b&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dlenta-ru-news.csv.gz&response-content-type=application%2Foctet-stream [following]
--2025-03-03 11:47:06--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/87156914/0b363e00-0126-11e9-9e3c-e8c235463bd6?X-Amz-Algorithm=AWS4-HMAC-

In [2]:
!pip install corus

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 83.7/83.7 kB 8.2 MB/s eta 0:00:00


In [4]:
!pip install natasha

  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 34.4/34.4 MB 9.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 46.7/46.7 kB 4.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 55.5/55.5 kB 6.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.2/8.2 MB 102.9 MB/s eta 0:00:00
  Created wheel for docopt: filename=docopt-0.6.2-py2.py3-none-any.whl size=13706 sha256=c2e39ae45dbc3591155001695d8f76a54c3acbc531802c666a3a94d85996477f
  Stored in directory: /root/.cache/pip/wheels/1a/b0/8c/4b75c4116c31f83c8f9f047231251e13cc74481cca4a78a9ce
  Created wheel for intervaltree: filename=intervaltree-3.1.0-py2.py3-none-any.whl size=26097 sha256=1618c9542fa365b79587245e22b7b49623bab08b22a5efbcd8ba860e6ee44781
  Stored in directory: /root/.cache/pip/wheels/31/d7/d9/eec6891f78cac19a693bd40ecb8365d2f4613318c145ec9816
Successfully built docopt intervaltree


In [26]:
from corus import load_lenta
import pandas as pd
import numpy as np
import random
import re
from natasha import Doc, Segmenter, MorphVocab
from sklearn.calibration import LabelEncoder
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.dummy import DummyClassifier
from sklearn.metrics import accuracy_score, classification_report
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.linear_model import LogisticRegression

**Обеспечим воспроизводимость ноутбука**

In [17]:
RANDOM_STATE = 42

def set_random_seed():
    np.random.seed(RANDOM_STATE)
    random.seed(RANDOM_STATE)

set_random_seed()

In [42]:
# Загрузка данных
path = 'lenta-ru-news.csv.gz'
records = load_lenta(path)
data = []

for record in records:
    if record.topic is None:
        continue
    data.append({
        'title': record.title,
        'text': record.text,
        'topic': record.topic
    })
    if len(data) >= 50000:
        break

df = pd.DataFrame(data)
class_counts = df['topic'].value_counts()
valid_classes = class_counts[class_counts > 500].index.to_list()
df = df[df['topic'].isin(valid_classes)]

In [43]:
df.head()

title  \
0  Названы регионы России с самой высокой смертно...   
1  Австрия не представила доказательств вины росс...   
2       Обнаружено самое счастливое место на планете   
3  В США раскрыли сумму расходов на расследование...   
4  Хакеры рассказали о планах Великобритании зами...   

                                                text        topic  
0  Вице-премьер по социальным вопросам Татьяна Го...       Россия  
1  Австрийские правоохранительные органы не предс...        Спорт  
2  Сотрудники социальной сети Instagram проанализ...  Путешествия  
3  С начала расследования российского вмешательст...          Мир  
4  Хакерская группировка Anonymous опубликовала н...          Мир

In [44]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 49337 entries, 0 to 49999
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   title   49337 non-null  object
 1   text    49337 non-null  object
 2   topic   49337 non-null  object
dtypes: object(3)
memory usage: 1.5+ MB


**Предобработку текста выполним с помощью регулярных выражений (это довольно быстро и удобно) и библиотеки natasha, так как она специально предназначена для обработки и анализа текстов на русском языке и отлично подойдет для нашей задачи:**
- Приведем всё к нижнему регистру
- Избавимся от чисел
- Удалим пунктуацию
- Удалим множественные пробелы
- Удалим URL-ссылки
- Избавимся от эмейлов
- Также выполним токенизацию и лемматизацию (она лучше стемминга, так как делает преобразования более аккуратно)

In [29]:
# Определим функцию нормализации и обработки текста
def normalize(text):
    if not isinstance(text, str) or pd.isna(text) or text == "":
        return ''

    try:
        text_lower = text.lower()
        clean_text = re.sub(r'\d+', '', text_lower)
        clean_text = re.sub(r'[^\w\s]', '', clean_text)
        clean_text = re.sub(r'\s+', ' ', clean_text).strip()
        clean_text = re.sub(r'https?://\S+|www\.\S+', '', clean_text)
        clean_text = re.sub(r'\S+@\S+', '', clean_text)
        doc = Doc(clean_text)
        doc.segment(segmenter)
        lemmatized_tokens = []

        for token in doc.tokens:
            try:
                token.lemmatize(morph_vocab)
                if hasattr(token, 'lemma') and token.lemma:
                    lemmatized_tokens.append(token.lemma)

                else:
                    lemmatized_tokens.append(token.text)

            except AttributeError:
                lemmatized_tokens.append(token.text)

        lemmatized_text = ' '.join(lemmatized_tokens)
        return lemmatized_text

    except Exception as e:
        print(f'Ошибка при обработке текста: {str(e)}')
        return clean_text

In [30]:
# Применим предобработку к текстам, заголовкам и топикам
morph_vocab = MorphVocab()
segmenter = Segmenter()

df['normalized_title'] = df['title'].apply(normalize)
df['normalized_text'] = df['text'].apply(normalize)
df['normalized_topic'] = df['topic'].apply(normalize)

In [31]:
# Разделим датасет на обучающую, валидационную и тестовую выборки со стратификацией
encoder = LabelEncoder()
dataset = df[['normalized_text', 'normalized_title', 'normalized_topic']]
X = dataset['normalized_title'] + ' ' + dataset['normalized_text']
y = encoder.fit_transform(dataset['normalized_topic'])

X_train, X_prep, y_train, y_prep  = train_test_split(X, y, test_size=0.4, random_state=42)
X_test, X_val, y_test, y_val = train_test_split(X_prep, y_prep, test_size=0.5, random_state=42)

In [32]:
# Замерим базовое качество с Dummy-бейзлайном
vectorizer = CountVectorizer(max_df=0.7, min_df=0.003)
X_train_vectorized = vectorizer.fit_transform(X_train)
X_val_vectorized = vectorizer.transform(X_val)
X_test_vectorized = vectorizer.transform(X_test)

dummy_clf = DummyClassifier(strategy="most_frequent").fit(X_train_vectorized, y_train)
preds = dummy_clf.predict(X_val_vectorized)
print(classification_report(y_val, preds))

              precision    recall  f1-score   support

           0       0.00      0.00      0.00       610
           1       0.00      0.00      0.00       408
           2       0.00      0.00      0.00       682
           3       0.00      0.00      0.00       772
           4       0.00      0.00      0.00       667
           5       0.14      1.00      0.24      1342
           6       0.00      0.00      0.00       701
           7       0.00      0.00      0.00       238
           8       0.00      0.00      0.00      1381
           9       0.00      0.00      0.00       536
          10       0.00      0.00      0.00      1129
          11       0.00      0.00      0.00       427
          12       0.00      0.00      0.00       975

    accuracy                           0.14      9868
   macro avg       0.01      0.08      0.02      9868
weighted avg       0.02      0.14      0.03      9868



/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


Пока что качество не очень, но оно и понятно, это ведь бейзлайн без какой-либо настройки

**Теперь обучим модель логистической регрессии с двумя вариантами векторизации:**
- CountVectorizer
- TfidfVectorizer

In [33]:
# CountVectorizer
logreg = LogisticRegression(max_iter=1000, random_state=42).fit(X_train_vectorized, y_train)
preds = logreg.predict(X_val_vectorized)
print(classification_report(y_val, preds))

              precision    recall  f1-score   support

           0       0.82      0.79      0.81       610
           1       0.89      0.83      0.86       408
           2       0.82      0.82      0.82       682
           3       0.79      0.79      0.79       772
           4       0.89      0.88      0.89       667
           5       0.82      0.85      0.83      1342
           6       0.86      0.86      0.86       701
           7       0.79      0.74      0.76       238
           8       0.75      0.78      0.77      1381
           9       0.78      0.76      0.77       536
          10       0.96      0.97      0.97      1129
          11       0.91      0.85      0.88       427
          12       0.85      0.83      0.84       975

    accuracy                           0.84      9868
   macro avg       0.84      0.83      0.83      9868
weighted avg       0.84      0.84      0.84      9868



In [35]:
# TfidfVectorizer
pipe_tfidf = Pipeline(
    steps=[
        ('tfidf', TfidfVectorizer()),
        ('clf', LogisticRegression(max_iter=1000, multi_class='ovr', random_state=42))
    ]
).fit(X_train, y_train)

preds = pipe_tfidf.predict(X_val)
print(classification_report(y_val, preds))

/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1256: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. Use OneVsRestClassifier(LogisticRegression(..)) instead. Leave it to its default value to avoid this warning.
  warnings.warn(


              precision    recall  f1-score   support

           0       0.85      0.82      0.84       610
           1       0.92      0.78      0.84       408
           2       0.80      0.84      0.82       682
           3       0.83      0.80      0.81       772
           4       0.90      0.90      0.90       667
           5       0.81      0.90      0.85      1342
           6       0.91      0.87      0.89       701
           7       0.89      0.55      0.68       238
           8       0.76      0.82      0.79      1381
           9       0.84      0.74      0.79       536
          10       0.95      0.98      0.96      1129
          11       0.95      0.81      0.88       427
          12       0.84      0.87      0.86       975

    accuracy                           0.85      9868
   macro avg       0.87      0.82      0.84      9868
weighted avg       0.85      0.85      0.85      9868



С применением TF-IDF векторизации результаты вышли чуточку лучше, поэтому будем использовать её

**Попробуем улучшить результаты работы нашей модели с помощью кросс-валидации и подбора оптимальных гиперпараметров. Будем варьировать размер n-грамм, размер словаря и коэффициент С для логистической регрессии. Как мне кажется, это наиболее важные параметры**

In [38]:
param_grid = {
    'tfidf__ngram_range': [(1, 1), (1, 2)],
    'tfidf__max_features': [10000, 20000],
    'clf__C': [0.1, 1, 10]
}

model = GridSearchCV(pipe_tfidf, param_grid, cv=3, n_jobs=-1)
model.fit(X_train, y_train)
print(f'Best params: {model.best_params_}')
print(f'Best CV accuracy: {model.best_score_}')

/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1256: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. Use OneVsRestClassifier(LogisticRegression(..)) instead. Leave it to its default value to avoid this warning.
  warnings.warn(


Best params: {'clf__C': 10, 'tfidf__max_features': 20000, 'tfidf__ngram_range': (1, 2)}
Best CV accuracy: 0.8657522909034162


In [41]:
best_model = model.best_estimator_
y_test_pred = best_model.predict(X_test)
print(f'Test accuracy: {accuracy_score(y_test, y_test_pred)}')
print()
print(classification_report(y_test, y_test_pred))

Test accuracy: 0.8732137427789601

              precision    recall  f1-score   support

           0       0.88      0.88      0.88       655
           1       0.90      0.87      0.89       397
           2       0.84      0.87      0.85       614
           3       0.86      0.80      0.83       767
           4       0.89      0.92      0.90       650
           5       0.85      0.89      0.87      1380
           6       0.93      0.90      0.92       738
           7       0.85      0.73      0.78       243
           8       0.81      0.83      0.82      1428
           9       0.82      0.83      0.83       539
          10       0.96      0.97      0.97      1029
          11       0.93      0.87      0.90       406
          12       0.87      0.87      0.87      1021

    accuracy                           0.87      9867
   macro avg       0.88      0.86      0.87      9867
weighted avg       0.87      0.87      0.87      9867



**Итоговое качество у нас получилось довольно высокое - Accuracy = 0.87. В целом, если увеличить сетку параметров и подождать подольше, то наверняка можно добиться и более высокого качества**